### PR TITLE
Simplify and centralise pause logic

### PR DESCRIPTION
**Pull request by peppy**
- Removes `PausableGameplayContainer`, in favour of logic being local to `Player`. The class felt quite redundant as it was basically just forwarding mes
sages between two other layers.
- `GameplayClockContainer` now controls the IsPaused variable.
- [x] Depends on #4485.

### CODE REVIEW

File: osu.Game/Screens/Play/Player.cs [hunk 1]
```diff
@@ -68,11 +66,33 @@ public class Player : ScreenWithBeatmapBackground
...
 
         protected HUDOverlay HUDOverlay { get; private set; }
-        private FailOverlay failOverlay;
+
+        #region Storyboard
 
         private DrawableStoryboard storyboard;
         protected UserDimContainer StoryboardContainer { get; private set; }
 
+        private void initializeStoryboard(bool asyncLoad)
```
Comment (smoogipoo) [inline osu.Game/Screens/Play/Player.cs]: Above `load()`?

File: osu.Game/Screens/Play/Player.cs [hunk 2]
```diff
@@ -350,18 +454,20 @@ public override bool OnExiting(IScreen next)
                 return true;
             }
 
-            if ((!AllowPause || HasFailed || !ValidForResume || PausableGameplayContainer?.IsPaused.Value != false || DrawableRuleset?.HasReplayLoaded.Value != false) && (!PausableGameplayContainer?.IsResuming ?? true))
+            if (LoadedBeatmapSuccessfully && canPause)
```
Comment (smoogipoo) [inline osu.Game/Screens/Play/Player.cs]:
```suggestion
            if (canPause)
```

`LoadedBeatmapSuccessfully` is a precondition in `canPause`.

File: osu.Game/Screens/Play/Player.cs [hunk 3]
```diff
@@ -289,19 +303,114 @@ protected virtual ScoreInfo CreateScore()
...
+
+        private bool canPause =>
+            // must pass basic screen conditions (beatmap loaded, instance allows pause)
+            LoadedBeatmapSuccessfully && AllowPause && ValidForResume
+            // replays cannot be paused and exit immediately
+            && !DrawableRuleset.HasReplayLoaded.Value
+            // cannot pause if we are already in a fail state
+            && !HasFailed
+            // cannot pause if already paused (and not in the process of resuming)
+            && (GameplayClockContainer.IsPaused.Value == false || IsResuming)
```
Comment (smoogipoo) [inline osu.Game/Screens/Play/Player.cs]: Isn't this supposed to be `!IsResuming`?

File: osu.Game/Screens/Play/Player.cs [hunk 4]
```diff
@@ -289,19 +303,114 @@ protected virtual ScoreInfo CreateScore()
...
+            && (!pauseCooldownActive || IsResuming);
```
Comment (smoogipoo) [inline osu.Game/Screens/Play/Player.cs]: I actually don't understand the separation between this line and the previous

**Review by smoogipoo — CHANGES_REQUESTED** [review]

File: osu.Game/Screens/Play/Player.cs (hunk 3 above)
Comment (peppy) [reply osu.Game/Screens/Play/Player.cs]: no, this is correct. in a state of resuming we need to allow re-pausing, as per the comment.

File: osu.Game/Screens/Play/Player.cs (hunk 4 above)
Comment (peppy) [reply osu.Game/Screens/Play/Player.cs]: neither do i

File: osu.Game/Screens/Play/Player.cs (hunk 3 above)
Comment (peppy) [reply osu.Game/Screens/Play/Player.cs]: i've reworded the comment to hopefully make this clearer

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**